File: ios-app/HAPhoneTestApp/EnvelopeVerifier.swift
import Foundation
import CryptoKit

/// Ed25519 envelope verification matching `tools/envelope.py` byte-for-byte.
///
/// Canonicalization contract (must stay identical to the Python reference):
///   - Only the fields in `canonicalFieldOrder` are ever serialized (the
///     `sig` field, if present, is always excluded).
///   - Keys are emitted in that fixed alphabetical order, no whitespace.
///   - Encoded as UTF-8 bytes before signing/verifying.
enum EnvelopeVerifier {
    static let canonicalFieldOrder = ["call_id", "call_type", "caller", "event_id", "expires_at", "issued_at", "v"]

    static func canonicalBytes(from dict: [String: Any]) -> Data? {
        var parts: [String] = []
        for key in canonicalFieldOrder {
            guard let value = dict[key] else { return nil }
            if let intVal = value as? Int {
                parts.append("\"\(key)\":\(intVal)")
            } else if let strVal = value as? String {
                let escaped = strVal
                    .replacingOccurrences(of: "\\", with: "\\\\")
                    .replacingOccurrences(of: "\"", with: "\\\"")
                parts.append("\"\(key)\":\"\(escaped)\"")
            } else {
                return nil
            }
        }
        return ("{" + parts.joined(separator: ",") + "}").data(using: .utf8)
    }

    static func verify(dict: [String: Any], publicKeyHex: String) -> Bool {
        guard let sigB64 = dict["sig"] as? String,
              let sigData = Data(base64Encoded: sigB64),
              let canonical = canonicalBytes(from: dict),
              let pubKeyData = Data(hexEncoded: publicKeyHex),
              let publicKey = try? Curve25519.Signing.PublicKey(rawRepresentation: pubKeyData)
        else { return false }
        return publicKey.isValidSignature(sigData, for: canonical)
    }

    static func isExpired(dict: [String: Any], now: Int = Int(Date().timeIntervalSince1970)) -> Bool {
        guard let expiresAt = dict["expires_at"] as? Int else { return true }
        return now > expiresAt
    }
}

extension Data {
    init?(hexEncoded hex: String) {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<next], radix: 16) else { return nil }
            data.append(byte)
            index = next
        }
        self = data
    }
}

File: app-flutter/android/app/src/main/kotlin/de/haphone/app/test/EnvelopeVerifier.kt
package de.haphone.app.test

import com.google.crypto.tink.subtle.Ed25519Verify
import java.util.Base64

object EnvelopeVerifier {
    val canonicalFieldOrder = listOf("call_id", "call_type", "caller", "event_id", "expires_at", "issued_at", "v")

    fun canonicalBytes(envelope: Map<String, Any>): ByteArray? {
        val parts = mutableListOf<String>()
        for (key in canonicalFieldOrder) {
            val value = envelope[key] ?: return null
            when (value) {
                is Int, is Long -> parts.add("\"$key\":$value")
                is String -> {
                    val escaped = value.replace("\\", "\\\\").replace("\"", "\\\"")
                    parts.add("\"$key\":\"$escaped\"")
                }
                else -> return null
            }
        }
        return ("{" + parts.joinToString(",") + "}").toByteArray(Charsets.UTF_8)
    }

    fun verify(envelope: Map<String, Any>, publicKeyHex: String): Boolean = try {
        val sigB64 = envelope["sig"] as? String ?: return false
        val sigBytes = Base64.getDecoder().decode(sigB64)
        val canonical = canonicalBytes(envelope) ?: return false
        val verifier = Ed25519Verify(hexToBytes(publicKeyHex))
        verifier.verify(sigBytes, canonical)
        true
    } catch (e: Exception) {
        false
    }

    fun isExpired(envelope: Map<String, Any>, now: Long = System.currentTimeMillis() / 1000): Boolean {
        val expiresAt = (envelope["expires_at"] as? Number)?.toLong() ?: return true
        return now > expiresAt
    }

    private fun hexToBytes(hex: String): ByteArray {
        val out = ByteArray(hex.length / 2)
        for (i in out.indices) {
            out[i] = ((Character.digit(hex[i * 2], 16) shl 4) + Character.digit(hex[i * 2 + 1], 16)).toByte()
        }
        return out
    }
}
